src/javascript/crypto/e2e/vrf/extra25519.js

```diff
@@ -0,0 +1,185 @@
+e2e.vrf.curve25519 = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.CURVE_25519);
+
+
+/**
+ * Unmarshals an elliptic curve point, and validates that it is an Ed25519 public key
+ * Refer to Definition 1 in https://www.iacr.org/archive/pkc2003/25670211/25670211.pdf
+ * 
+ * @param {!e2e.ByteArray} sBytes A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.extra25519.fromBytesBaseGroup = function(sBytes) {
+
+	var P = e2e.vrf.ed25519.curve.pointFromByteArray(sBytes);
+
+	goog.asserts.assert(!P.isInfinity(),
+	// goog.asserts.assert(!P.isIdentity(),
+        'P != infinity');
+
+	goog.asserts.assert(P.isOnCurve(),
+        'P must satisfy the defining equation of the elliptic curve E');
+
+	// TODO: confirm https://github.com/google/end-to-end/issues/340
+	
+	// nP = infinity, where n is the order 
+	// goog.asserts.assert(P.multiply(e2e.vrf.ed25519.n).isInfinity(), 
+	// goog.asserts.assert(P.multiply(e2e.vrf.ed25519.n).isIdentity(),
+		// 'nP = infinity, where n is the order');
+
+	return P;
+}
+
+/**
+ * Returns a*A + b*B, where B is the Ed25519 base point
+ *
+ * @param {!e2e.BigNum} a the representative coordinate
+ * @param {!e2e.ecc.Element} A the representative coordinate
+ * @param {!e2e.BigNum} b the representative coordinate
+ * @return {!e2e.ecc.Element}
+ */
+e2e.vrf.extra25519.geDoubleScalarMultVartime = function(a, A, b) {
+	return A.multiply(a).add(e2e.vrf.ed25519.curve.B.multiply(b));
+}
+
+
+// converts a uniform representative value for a
+// curve25519 public key, as produced by ScalarBaseMult, to a curve25519 public
+// key.
+// See Elligator paper section 5.2.
+
+/**
+ * Converts a uniform representative value to an element in Curve25519, 
+ *  using the elligator map
+ * Returns an element that is either -A/(1+ur^2) or A - A/(1+ur^2), 
+ *  where u = 2, A = 486662 (defined in curve25519)
+ *
+ * Refer to Section 5.2 in http://elligator.cr.yp.to/elligator-20130828.pdf
+ * @param {!e2e.ecc.Element} r the representative coordinate
+ * @return {!e2e.ecc.Element}
+ */
+function representativeToMontgomeryX(r) {
+	var A = e2e.vrf.curve25519.curve.A;
+
+	// v = -A/(1+ur^2), where u = 2 (any non-square field element)
+	r = r.square();
+	var v = r.add(r).add(e2e.vrf.ed25519.curve.ONE).inverse().multiply(A).negate();
+
+	// assertEquals('d4ad43e1aaf9b0ce31093a2cbe62af7e53bcb072c804e23b0d395147be6eed44',
+	// 	goog.crypt.byteArrayToHex(v.x.toByteArray().reverse()));
+
+	// e = Chi(v^3 + Av^2 + Bv), where B = 1
+	// Chi(z) = z ^ (q - 1) / 2
+	var v2 = v.square();
+	var v3 = v2.multiply(v);
+	var e = v3.add(v2.multiply(A)).add(v)
+			.power(e2e.vrf.ed25519.curve.q.subtract(e2e.BigNum.ONE).shiftRight(1));
+
+	// leading zeros are dropped
+	// assertEquals('01',
+	// 	goog.crypt.byteArrayToHex(e.x.toByteArray().reverse()));
+
+	// e is either 1 or -1
+	// x = ev - (1-e)A/2
+	var x = e.isEqual(e2e.vrf.ed25519.curve.ONE) ? v : A.substract(v);
```

Comment:
Please document that this is not constant time and thus must not be used with secret inputs.